src/DotNetty.Codecs/DelimiterBasedFrameDecoder.cs

```diff
@@ -0,0 +1,286 @@
+﻿using System;
+using System.Collections.Generic;
+using DotNetty.Buffers;
+using DotNetty.Transport.Channels;
+
+namespace DotNetty.Codecs
+{
+    /// <summary>
+    /// A decoder that splits the received <see cref="DotNetty.Buffers.IByteBuffer"/> by one or more
+    /// delimiters.It is particularly useful for decoding the frames which ends
+    /// with a delimiter such as <see cref = "DotNetty.Codecs.Delimiters.NullDelimiter" /> or
+    /// <see cref = "DotNetty.Codecs.Delimiters.LineDelimiter" />
+    /// --- Specifying more than one delimiter </ h3 >
+    /// <see cref = "DotNetty.Codecs.Delimiters.NullDelimiter" /> allows you to specify more than one
+    /// delimiter.  If more than one delimiter is found in the buffer, it chooses
+    /// the delimiter which produces the shortest frame.  For example, if you have
+    /// the following data in the buffer:
+    ///  +--------------+
+    ///  | ABC\nDEF\r\n |
+    ///  +--------------+
+    ///  a <see cref = "DotNetty.Codecs.Delimiters.LineDelimiter" /> will choose '\n' as the first delimiter and produce two frames:
+    ///  +-----+-----+
+    ///   | ABC | DEF |
+    ///  +-----+-----+
+    ///  rather than incorrectly choosing '\r\n' as the first delimiter:
+    ///  +----------+
+    ///  | ABC\nDEF |
+    ///  +----------+
+    /// </summary>
+    public class DelimiterBasedFrameDecoder : ByteToMessageDecoder
+    {
+        private IByteBuffer[] delimiters;
```

Comment:
code style: do not specify default visibility (private on type members, internal on types)